src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+
+    meta_type = "CompositeIndex"
+
+    manage_options = (
+        {'label': 'Settings',
+         'action': 'manage_main',
+         'help': ('CompositeIndex', 'CompositeIndex_Settings.stx')},
+        {'label': 'Browse',
+         'action': 'manage_browse',
+         'help': ('CompositeIndex', 'CompositeIndex_Settings.stx')},
+    )
+
+    query_options = ("query", "operator")
+
+    def __init__(self, id, ignore_ex=None, call_methods=None,
+                 extra=None, caller=None):
+        """Create an composite index
+
+        UnIndexes are indexes that contain two index components, the
+        forward index (like plain index objects) and an inverted
+        index.  The inverted index is so that objects can be unindexed
+        even when the old value of the object is not known.
+
+        e.g.
+
+        self._index = {datum:[documentId1, documentId2]}
+        self._unindex = {documentId:datum}
+
+        If any item in self._index has a length-one value, the value is an
+        integer, and not a set.  There are special cases in the code to deal
+        with this.
+
+        The arguments are:
+
+          'id' -- the name of the item attribute to index.  This is
+          either an attribute name or a record key.
+
+          'ignore_ex' -- should be set to true if you want the index
+          to ignore exceptions raised while indexing instead of
+          propagating them.
+
+          'call_methods' -- should be set to true if you want the index
+          to call the attribute 'id' (note: 'id' should be callable!)
+          You will also need to pass in an object in the index and
+          uninded methods for this to work.
+
+          'extra' -- a mapping object that keeps additional
+          index-related parameters - subitem 'indexed_attrs'
+          can be list of dicts with following keys { id, type, attributes }
+
+          'caller' -- reference to the calling object (usually
+          a (Z)Catalog instance
+        """
+
+        self.id = id
+        self.ignore_ex = ignore_ex        # currently unimplimented
+        self.call_methods = call_methods
+
+        self.operators = ('or', 'and')
+        self.useOperator = 'or'
+
+        # set components
+        self._components = ComponentMapping()
+        if extra:
+            for cdata in extra:
+                c_id = cdata['id']
+                c_meta_type = cdata['meta_type']
+                c_attributes = cdata['attributes']
+                self._components[c_id] = Component(c_id, c_meta_type,
+                                                   c_attributes)
+        self.clear()
+
+    def _apply_index(self, request, resultset=None):
+        """ Apply the index to query parameters given in the request arg. """
+        record = parseIndexRequest(request, self.id, self.query_options)
+
+        if record.keys is None:
+            return None
+
+        return UnIndex._apply_index(self, request, resultset=resultset)
```

Comment:
_apply_index looks like it doesn't actually do anything different, compared to UnIndex._apply_index. Did you mean to add anything here? Otherwise you could just remove it here and directly rely on the inherited method.